{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6d885d4c-4502-4c58-ba06-f25e09f8605d",
   "metadata": {},
   "source": [
    "# Part II: Get metadata\n",
    "\n",
    "- Number of transcripts per cell and per gene\n",
    "- Relate original cell ID with `ndimage` one\n",
    "- Some other data related to cell and overall concentration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "956e43c6-4044-4db4-b9ff-ba4343889c59",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import tifffile as tf\n",
    "from glob import glob\n",
    "import os\n",
    "from scipy import ndimage, interpolate, spatial, stats\n",
    "\n",
    "from importlib import reload\n",
    "import utils"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e1fa818-1b3c-43e8-8d9a-aaa1eb4bb29f",
   "metadata": {},
   "source": [
    "# Load all general data\n",
    "\n",
    "- Same setup as in Part I"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c0d39d91-b5dd-4c6d-af05-06e6962d7b89",
   "metadata": {},
   "outputs": [],
   "source": [
    "wsrc = '../cell_dams/'\n",
    "nsrc = '../nuclear_mask/'\n",
    "tsrc = '../translocs/'\n",
    "psrc = '../proc/'\n",
    "osrc = '../data/'\n",
    "sample = 'D2'\n",
    "\n",
    "dst = '../kde/'\n",
    "\n",
    "dst += sample + os.sep\n",
    "if not os.path.isdir(dst):\n",
    "    os.mkdir(dst)\n",
    "\n",
    "wall = tf.imread(wsrc + sample + '_dams.tif').astype(bool)\n",
    "nuclei = tf.imread(nsrc + sample + '_EDT.tif') < 2\n",
    "label, cellnum = ndimage.label(wall, ndimage.generate_binary_structure(2,1))\n",
    "print('Detected',cellnum,'cells')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17365b71-7b91-423b-b1bb-fdd3499818dc",
   "metadata": {},
   "source": [
    "- Load all the CSVs with spatial locations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "ec0a5878-69a8-4c84-8ace-c407de465af5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "filenames = sorted(glob(tsrc + sample + os.sep + '*.csv'))\n",
    "tsize = np.zeros(len(filenames), dtype=int)\n",
    "transcriptomes = [os.path.splitext(filenames[i])[0].split('_-_')[-1] for i in range(len(filenames)) ]\n",
    "translocs = [None for i in range(len(filenames))]\n",
    "for i in range(len(filenames)):\n",
    "    translocs[i] = pd.read_csv(filenames[i], header=None, names=['X', 'Y', 'Z'])\n",
    "    tsize[i] = len(translocs[i])\n",
    "tlocs = pd.concat(translocs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95e2b6ff-8d83-43c5-84cb-cf46cbd06187",
   "metadata": {},
   "source": [
    "# Compute metadata\n",
    "\n",
    "Get metadata for every cell\n",
    "-  Lower-left and upper right corners of the minimal bounding box containing the cell\n",
    "-  Length and height of such box\n",
    "-  Total cell area\n",
    "-  Total area occupied by nuclei\n",
    "-  Total area occupied by cytosol\n",
    "-  Cytosol-to-total area ratio\n",
    "-  Center of mass according to the original data\n",
    "-  Center of mass according to `ndimage` computed right now\n",
    "-  Cell ID from the original data\n",
    "-  Label ID matched from `ndimage` (matching done by comparing centers of mass)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "07bbee5d-5ea7-4650-a2f9-aa22bb895b91",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../kde/D2/D2_cells_metadata.csv\n",
      "(2937, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>length</th>\n",
       "      <td>2937</td>\n",
       "      <td>309</td>\n",
       "      <td>336</td>\n",
       "      <td>77</td>\n",
       "      <td>209</td>\n",
       "      <td>278</td>\n",
       "      <td>372</td>\n",
       "      <td>17152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>height</th>\n",
       "      <td>2937</td>\n",
       "      <td>309</td>\n",
       "      <td>338</td>\n",
       "      <td>81</td>\n",
       "      <td>205</td>\n",
       "      <td>278</td>\n",
       "      <td>379</td>\n",
       "      <td>17152</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        count  mean  std  min  25%  50%  75%    max\n",
       "length   2937   309  336   77  209  278  372  17152\n",
       "height   2937   309  338   81  205  278  379  17152"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reload(utils)\n",
    "filename = dst + sample + '_cells_metadata.csv'\n",
    "print(filename)\n",
    "if not os.path.isfile(filename):\n",
    "    celllocs = utils.celllocs_read(osrc + sample + '_data/' + transcriptomes[1] + '/' + transcriptomes[1] + ' - localization results by cell.csv')\n",
    "    dcoords, cnuclei, cmatches = utils.match_original_ndimage(celllocs, wall, label, cellnum)\n",
    "    objss = ndimage.find_objects(label)\n",
    "    meta = utils.generate_cell_metadata(label, objss, nuclei)\n",
    "    meta = meta.join(pd.DataFrame(np.round(dcoords[cmatches], 2), columns=['orig_comX', 'orig_comY']))\n",
    "    meta = meta.join(pd.DataFrame(np.round(np.flip(cnuclei, axis=1),2), columns=['ndimage_comX', 'ndimage_comY']))\n",
    "    meta['orig_cellID'] = celllocs['Cell.ID..'].values[cmatches]\n",
    "    meta['ndimage_cellID'] = np.arange(1,cellnum+1)\n",
    "    meta.to_csv(filename, index=False)\n",
    "\n",
    "metacell = pd.read_csv(filename)\n",
    "print(metacell.shape)\n",
    "metacell.loc[:,['length', 'height']].describe().astype(int).T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11da59b0-bdc8-4dab-9e76-47620acf254a",
   "metadata": {},
   "source": [
    "Get metadata for every gene\n",
    "\n",
    "- Total number of transcripts throughout the cross section\n",
    "- Total number of transcripts floating in cytosol\n",
    "- Total number floating in nuclei\n",
    "- Cytosolic-to-total transcript ratio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "0f7cafb1-9e56-4713-b885-b0d4c2a9e9fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../kde/D2/D2_transcripts_metadata.csv\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total_number</th>\n",
       "      <th>cyto_number</th>\n",
       "      <th>nuclei_number</th>\n",
       "      <th>ratio</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>32613</td>\n",
       "      <td>32171</td>\n",
       "      <td>442</td>\n",
       "      <td>0.986447</td>\n",
       "      <td>BAC45727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1162</td>\n",
       "      <td>1151</td>\n",
       "      <td>11</td>\n",
       "      <td>0.990534</td>\n",
       "      <td>BAC46169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1820064</td>\n",
       "      <td>1795069</td>\n",
       "      <td>24995</td>\n",
       "      <td>0.986267</td>\n",
       "      <td>BAC47034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1939</td>\n",
       "      <td>1923</td>\n",
       "      <td>16</td>\n",
       "      <td>0.991748</td>\n",
       "      <td>BAC47856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3033</td>\n",
       "      <td>2985</td>\n",
       "      <td>48</td>\n",
       "      <td>0.984174</td>\n",
       "      <td>BAC48395</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   total_number  cyto_number  nuclei_number     ratio      gene\n",
       "0         32613        32171            442  0.986447  BAC45727\n",
       "1          1162         1151             11  0.990534  BAC46169\n",
       "2       1820064      1795069          24995  0.986267  BAC47034\n",
       "3          1939         1923             16  0.991748  BAC47856\n",
       "4          3033         2985             48  0.984174  BAC48395"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filename = dst + sample + '_transcripts_metadata.csv'\n",
    "print(filename)\n",
    "if not os.path.isfile(filename):\n",
    "    data = pd.read_csv(osrc + sample + '_data' + os.sep + '32771-slide1_' + sample + '_results.txt', header=None, sep='\\t').drop(columns=[4])\n",
    "    _, orig_size = np.unique(data.iloc[:,-1], return_index = False, return_inverse=False, return_counts=True) \n",
    "    meta = pd.DataFrame()\n",
    "    meta['total_number'] = orig_size\n",
    "    meta['cyto_number'] = tsize\n",
    "    meta['nuclei_number'] = orig_size - tsize\n",
    "    meta['ratio'] = tsize/orig_size\n",
    "    meta['gene'] = transcriptomes\n",
    "    \n",
    "    meta.to_csv(filename, index=False)\n",
    "\n",
    "metatrans = pd.read_csv(filename)\n",
    "metatrans.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20e97625-6b40-4738-a1c3-6f0124d0be67",
   "metadata": {},
   "source": [
    "Get metadata for every cell and gene\n",
    "\n",
    "- Number of transcripts of a given gene in a given cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a07e06e6-b398-4882-a913-95912245ac06",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../kde/D2/D2_transcells_metadata.csv\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>...</th>\n",
       "      <th>2929</th>\n",
       "      <th>2930</th>\n",
       "      <th>2931</th>\n",
       "      <th>2932</th>\n",
       "      <th>2933</th>\n",
       "      <th>2934</th>\n",
       "      <th>2935</th>\n",
       "      <th>2936</th>\n",
       "      <th>2937</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>BAC45727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>BAC46169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>859</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>BAC47034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>BAC47856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>BAC48395</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 2938 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     1  2  3  4  5  6  7  8  9  10  ...  2929  2930  2931  2932  2933  2934  \\\n",
       "0   31  0  0  0  0  0  0  0  0   0  ...     0     0     0     0     0     0   \n",
       "1    7  0  0  0  0  0  0  0  1   0  ...     0     0     0     0     0     0   \n",
       "2  859  1  0  1  1  0  0  0  0   0  ...     0     1     0     1     0     2   \n",
       "3    2  0  0  0  0  0  0  0  0   0  ...     0     0     0     0     0     0   \n",
       "4    5  0  0  0  0  0  0  0  0   0  ...     0     0     0     0     0     0   \n",
       "\n",
       "   2935  2936  2937      gene  \n",
       "0     0     0     0  BAC45727  \n",
       "1     0     0     0  BAC46169  \n",
       "2     0     0     0  BAC47034  \n",
       "3     0     0     0  BAC47856  \n",
       "4     0     0     0  BAC48395  \n",
       "\n",
       "[5 rows x 2938 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filename = dst + sample + '_transcells_metadata.csv'\n",
    "print(filename)\n",
    "if not os.path.isfile(filename):\n",
    "    meta = utils.generate_transcell_metadata(translocs, transcriptomes, cellnum, label)\n",
    "    meta.to_csv(filename, index=False)\n",
    "transcell = pd.read_csv(filename)\n",
    "transcell.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
